---
title: "Bioinformatic Read and Write Sequences"
author: "Truc Huynh"
date: "2/11/2021"
output:
  html_document: default
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
library(seqinr)
```

## Objectives
- Explore features of the Bioconductor Package
- Query/read/write sequence data.

## Description

Write a R code to do each of the following tasks :

### Question a:
  Search for DNA sequences from the organism “Chlamydia trachomatis” in
the ACNUC “genbank”.
```{r DNA_Search}
choosebank("genbank")
Q <- query("Q", "SP=Chlamydia trachomatis") 

```

### Question b: 
  How many sequences retrieved?
```{r sequences_retrieved}

Q$name
Q$nelem
```
- 
  
  
### Question c: 
  How many bases are there in longest sequence among them?
```{r longest_Sequences}
max(sapply(Q$req, getLength))

```

### Question d: 
  For the first three sequences, print out the accession numbers?
```{r accession numbers}
getName(Q$req[1])
getName(Q$req[2])
getName(Q$req[3])

```

### Question e: 
  For the 1000th sequence, print out the nucleotide bases in the range 50 to 75.
```{r printNuBase}
s1000 = getSequence(Q$req[[1000]])

#Print the first 10 bases in the sequence
s1000[50:75]
```


### Question f: 
  What is the length of the 250th sequence?
```{r length}

# length Vs. getLength in R
# getLength is use to get the length of a sequence in R
getLength(Q$req[[250]])
```

175 is the length of 250th sequences

### Question g: 
  Export the 150th , 151th, and 152th sequences into a FASTA file.
  The write.fasta() function accept an input as vector of sequences and vector of sequences name and allocate the name with appropriate sequences.
  
```{r export}
VectorSeq <- c(getSequence(Q$req[150]),getSequence(Q$req[151]),getSequence(Q$req[152]))
VectorSeqName <- c(getName(Q$req[150]),getName(Q$req[151]),getName(Q$req[152]))
write.fasta(VectorSeq, VectorSeqName, file.out = "Sequences.fasta")
```

```{r close}
closebank()
```
## Notes:

• Handwritten answers are not allowed!
• Use Rmarkdown (https://rmarkdown.rstudio.com/) and provide a neatly formatted
“pdf” file showing both code and output.
• Include your name as a comment at the beginning of the script file.