---
title: "Sequence_Aglinment"
author: "Truc Huynh"
date: "3/12/2021"
output:
  word_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
library(seqinr)
library(Biostrings)
```

## Objectives 
- Query/read/Analyze DNA sequence data.
- Create and use score matrices.
- Score and analyze sequence alignments

### Question 1:
Retrieve the 2 sequences "AY884001" and "MH940245" from “genbank”.

```{r sequenceSARS}
# Choose the ACNUC
choosebank("genbank")

# Retrieve sequence AY884001
My_Que1 <- query("My_Que1", "AC=AY884001")

# write  sequence AY884001 to fasta file
write.fasta(getSequence(My_Que1[['req']][[1]]),getName(My_Que1[['req']][[1]]),"AY88.fasta")

Seq1 <- (getSequence(read.fasta("AY88.fasta")))

# Retrieve sequence "MH940245"
My_Que2 <- query("My_Que2", "AC=MH940245")

# write sequence MH940245 to fasta
write.fasta(getSequence(My_Que2[['req']][[1]]), getName(My_Que2[['req']][[1]]),"MH94.fasta")

Seq2 <- getSequence(read.fasta("MH94.fasta"))

closebank()

```


### Question 2:
for each sequence, compute the frequency of each amino acid and plot
them as a pie chart. Generate only one figure for both sequences.


```{r frequency}
# sequences AY884001
t1 = seqinr::count(getSequence(Seq1[[1]]), wordsize = 3)
print("frequency of each amino acid in sequences AY884001")
t1

# sequence MH940245
t2= count(getSequence(Seq2[[1]]), wordsize = 3)
print("frequency of each amino acid in sequences AY884001")
t2 

# Check if they are equal
t1==t2

# Draw Pie Chart
old.par <- par(mfrow=c(1, 2))
pie(t1,main="AY884001")
pie(t2,main="MH940245")
par(old.par)

```

### Question 3:
Create a dot plot for the first ORF in each of the sequences. Comment on 
the result.
```{r dotPlot}
dotPlot(Seq1[[1]], Seq2[[1]], wsize = 3, wstep = 3, nmatch = 3)
```

### Question 4:
Find the optimal global alignment between the two sequences and print 
the alignment for the first 20 nucleotides. Use a score +2 for a match, -1 for a 
mismatch, and the gap penalty = 2.
```{r transform}
#Transform to Upper Case
Seq1 <- toupper(c2s(Seq1[[1]]))
Seq2 <- toupper(c2s(Seq2[[1]]))
```

```{r findOptimal}
sigma <- nucleotideSubstitutionMatrix(match = 2, mismatch = -1, baseOnly = TRUE)
sigma # Print out the matrix

# Optimal Global Aglinment
pairwiseSeq1Seq2<-pairwiseAlignment(Seq1, Seq2, substitutionMatrix = sigma, gapOpening = 2, scoreOnly = FALSE)
```
```{r displayOptimal}
pairwiseSeq1Seq2
pairwiseSeq1Seq2@score
```
### Question 5:
Is the global alignment statistically significant? Explain your answer

```{r generateSeqsWithMultinomialModel}
generateSeqsWithMultinomialModel <- function(inputsequence, n)
{
  # Change the input sequence into a vector of letters
    require("seqinr") # This function requires the SeqinR package.
    inputsequencevector <- s2c(inputsequence)
    #inputsequencevector <- inputsequence
    # Find the frequencies of the letters in the input sequence "inputsequencevector":
    mylength <- length(inputsequencevector)
    mytable <- table(inputsequencevector)
    # Find the names of the letters in the sequence
    letters <- rownames(mytable)
    numletters <- length(letters)
    probabilities <- numeric() # Make a vector to store the probabilities of letters
    for (i in 1:numletters)
    {
      letter <- letters[i]
      count <- mytable[[i]]
      probabilities[i] <- count/mylength
    }
    # Make n random sequences using the multinomial model with probabilities "probabilities"
    seqs <- vector("list", n)
    for (j in 1:n)
    {
      seq <- sample(letters, mylength, rep=TRUE, prob=probabilities) # Sample
      seq <- c2s(seq)
      seqs[[j]] <- seq
    }
    
    
# Return the vector of random sequences
return(seqs)
}
```
- Create a vector of 1000 random sequences.
```{r randomSeq}
randomSeq <- generateSeqsWithMultinomialModel(Seq2,1000)

```
- Use PairwiseAlignment to get the score vector of 1000 random vector and store in randomScore

```{r randomScore}
randomscores <- double(1000)

for (i in 1:1000)
{
  score <- pairwiseAlignment(Seq1, randomSeq[[i]], substitutionMatrix = sigma, gapOpening = 2, scoreOnly = TRUE)
  randomscores[i] <- score
}
```

```{r Pvalue}
Pvalue<- sum(randomscores >= pairwiseSeq1Seq2@score)/1000
Pvalue
```

### Question 6:
What is the score of the optimal local alignment between the two sequences?
What is the length of the aligned segments? Use a score +3 for a match,-2 for a mismatch, the gap penalty = 4, and gap extension = 2.
```{r findOptimal2}
sigma <- nucleotideSubstitutionMatrix(match = 3, mismatch = -2, baseOnly = TRUE)
sigma # Print out the matrix

pairwiseAlignment(Seq1, Seq2, substitutionMatrix = sigma, gapOpening = 4, gapExtension= 2,  scoreOnly = FALSE)
```