---
title: "HW9-TrucHuynh"
author: "Truc Huynh"
date: "4/16/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
library(seqinr)
library(ape)
source("rootedNJtree.R")
source("unrootedNJtree.R")
source("cleanAlignment.R")
```

## Objectives
- Query and read list of sequence data.
- Produce and analyze multiple sequence alignments.
- Construct, plot, and analyze Phylogenetic trees.

### Question 1:
Retrieve/download the following DNA sequences from GenBank:
SARS coronavirus MA15 ExoN1 (FJ882953)
SARS coronavirus Frankfurt (AB257344)
SARS coronavirus ZJ0301 (DQ182595) 
SARS coronavirus strain CV7 (DQ898174)
SARS-CoV-2 (MT334547)

I was encounter this issues while retrieved "genbank", so I directly download the sequences and paste them in a FASTA file instead of retrieving 
Error: Error in choosebank("genbank") : I wasn't able to open the socket connection: o Check that your are connected to the internet. o Check that port 5558 is not closed by a firewall. o Try to increase timeout value (current is 5 seconds).

```{r createData}
# Choose the ACNUC


```
### Question 2:
Build and plot an unrooted phylogenetic tree from the sequence data using the neighbor-joining algorithm. For better display results you may set the type of the phylogenetic tree using type=phylogram



```{r readAln}
# Read in Multiple Alignment Sequences
objAln<-read.alignment(file="virus.phy", format = 'phylip')
unrootedNJtree(objAln,'protein')

```

```{r cleanAln}
objAlnClean<-cleanAlignment(objAln,30,30)
unrootedNJtree(objAlnClean,'protein')
```

```{r unrooted}
objAln<-read.alignment(file="virus2.phy", format = 'phylip')

dist.alignment(objAln)
objAln$nam[6]

treeTruc<-rootedNJtree(objAln,objAln$nam[6],'protein')
write.tree(treeTruc,"file.tre")
treeTruc
```

